Userland/Libraries/LibCrypto/Hash/BLAKE2b.cpp

```diff
@@ -0,0 +1,129 @@
+        if (inlen > fill) {
+            m_internal_state.buffer_length = 0;
+            // Fill the buffer.
+            __builtin_memcpy(m_internal_state.buffer + left, in, fill);
+
+            increment_counter_by(BLAKE2bConstants::blockbytes);
+            transform(m_internal_state.buffer);
+            in += fill;
+            inlen -= fill;
+            while (inlen > BLAKE2bConstants::blockbytes) {
+                increment_counter_by(BLAKE2bConstants::blockbytes);
+                transform(in);
+                in += BLAKE2bConstants::blockbytes;
+                inlen -= BLAKE2bConstants::blockbytes;
+            }
+        }
+        __builtin_memcpy(m_internal_state.buffer + m_internal_state.buffer_length, in, inlen);
+        m_internal_state.buffer_length += inlen;
+    }
+}
+
+BLAKE2b::DigestType BLAKE2b::peek()
+{
+    DigestType digest;
+    increment_counter_by(m_internal_state.buffer_length);
+
+    if (m_internal_state.last_node)
+        m_internal_state.is_at_last_node = UINT64_MAX;
+
+    // Set this as the last block
+    m_internal_state.is_at_last_block = UINT64_MAX;
+
+    // Pad the buffer with zeros
+    __builtin_memset(m_internal_state.buffer + m_internal_state.buffer_length, 0, BLAKE2bConstants::blockbytes - m_internal_state.buffer_length);
+    transform(m_internal_state.buffer);
+
+    for (size_t i = 0; i < 8; ++i)
+        __builtin_memcpy(&digest.data[0] + sizeof(m_internal_state.hash_state[i]) * i, &m_internal_state.hash_state[i], sizeof(m_internal_state.hash_state[i]));
+
+    return digest;
+}
+
+BLAKE2b::DigestType BLAKE2b::digest()
+{
+    auto digest = peek();
+    reset();
+    return digest;
+}
+
+void BLAKE2b::increment_counter_by(const u64 amount)
+{
+    m_internal_state.message_byte_offset[0] += amount;
+    m_internal_state.message_byte_offset[1] += (m_internal_state.message_byte_offset[0] < amount);
+}
+
+void BLAKE2b::mix(u64* work_array, u64 a, u64 b, u64 c, u64 d, u64 x, u64 y)
+{
+    constexpr auto rotation_constant_1 = 32;
+    constexpr auto rotation_constant_2 = 24;
+    constexpr auto rotation_constant_3 = 16;
+    constexpr auto rotation_constant_4 = 63;
+
+    work_array[a] = work_array[a] + work_array[b] + x;
+    work_array[d] = ROTRIGHT(work_array[d] ^ work_array[a], rotation_constant_1);
+    work_array[c] = work_array[c] + work_array[d];
+    work_array[b] = ROTRIGHT(work_array[b] ^ work_array[c], rotation_constant_2);
+    work_array[a] = work_array[a] + work_array[b] + y;
+    work_array[d] = ROTRIGHT(work_array[d] ^ work_array[a], rotation_constant_3);
+    work_array[c] = work_array[c] + work_array[d];
+    work_array[b] = ROTRIGHT(work_array[b] ^ work_array[c], rotation_constant_4);
+}
+
+void BLAKE2b::transform(u8 const* block)
+{
+    u64 m[16];
+    u64 v[16];
+
+    for (size_t i = 0; i < 16; ++i)
+        // FIXME: This only works on little-endian architectures.
+        m[i] = *reinterpret_cast<u64 const*>(block + i * sizeof(m[i]));
```

Comment:
This casting business is no good, not only is it Bad For Portability(:tm:), the alignment requirements are also not necessarily met here.
We have `ByteReader::load()` for this, use that. Expect a ~5% perf hit on x86.